ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala

```diff
@@ -10,19 +10,24 @@ import io.grpc.stub.StreamObserver
 import scala.concurrent.duration.FiniteDuration
 import scala.concurrent.{ExecutionContext, Future, Promise}
 
-final class TimeBoundObserver[T](duration: FiniteDuration)(
+final class TimeBoundObserver[T](duration: FiniteDuration, maximumResults: Option[Int] = None)(
     implicit executionContext: ExecutionContext)
     extends StreamObserver[T] {
 
   private val promise = Promise[Vector[T]]
   private val buffer = Vector.newBuilder[T]
+  private var size = 0
 
   Delayed.by(duration)(onCompleted())
 
   def result: Future[Vector[T]] = promise.future
 
   override def onNext(value: T): Unit = {
     buffer += value
+    size += 1
+    if (maximumResults.exists(size >= _)) {
+      onCompleted()
```

Comment:
This could possibly fire concurrently with the timer.

1. Is `onCompleted` reentrant?
2. Would it make sense to gate both `onNext` and `onCompleted` on a shared lock (e.g. `promise.synchronized`) and check that the promise is not completed when `onNext` enters the critical section?

Reply:
Shouldn't be a problem; `onCompleted` will work idempotently. This logic was copied from `SizeBoundObserver`.

I want to refactor these observers so they're decorators which can be layered on top of each other, but it seems like a big change, so I'll do it separately after this is in.

Reply:
Addressed in https://github.com/digital-asset/daml/pull/7984.